src/flat.rs

```diff
@@ -0,0 +1,1139 @@
+    /// `index_of(1, 2) = 4`.
+    pub fn min_length(&self) -> Option<usize> {
+        self.format.min_length()
+    }
+
+    /// Check if the buffer is large enough.
+    pub fn fits(&self, len: usize) -> bool {
+        self.format.fits(len)
+    }
+
+    /// If there are any samples aliasing each other.
+    ///
+    /// If this is not the case, it would always be safe to allow mutable access to two different
+    /// samples at the same time. Otherwise, this operation would need additional checks. When one
+    /// dimension overflows `usize` with its stride we also consider this aliasing.
+    pub fn has_aliased_samples(&self) -> bool {
+        self.format.has_aliased_samples()
+    }
+
+    /// Check if a buffer fulfills the requirements of a normal form.
+    ///
+    /// Certain conversions have preconditions on the structure of the sample buffer that are not
+    /// captured (by design) by the type system. These are then checked before the conversion. Such
+    /// checks can all be done in constant time and will not inspect the buffer content. You can
+    /// perform these checks yourself when the conversion is not required at this moment but maybe
+    /// still performed later.
+    pub fn is_normal(&self, form: NormalForm) -> bool {
+        self.format.is_normal(form)
+    }
+
+    /// Check that the pixel and the channel index are in bounds.
+    pub fn in_bounds(&self, x: u32, y: u32, channel: u8) -> bool {
+        self.format.in_bounds(x, y, channel)
+    }
+
+    /// Resolve the index of a particular sample.
+    ///
+    /// `None` if the index is outside the bounds or does not fit into a `usize`.
+    pub fn index(&self, x: u32, y: u32, channel: u8) -> Option<usize> {
+        self.format.index(x, y, channel)
+    }
+
+    /// Get the theoretical position of sample (x, y, channel).
+    ///
+    /// The 'check' is for overflow during index calculation, not that it is contained in the
+    /// image.
+    pub fn index_ignoring_bounds(&self, x: usize, y: usize, channel: usize) -> Option<usize> {
+        self.format.index_ignoring_bounds(x, y, channel)
+    }
+
+    /// Get an index provided it is inbouds.
+    ///
+    /// The computation is assumed to not overflow. It is unspecified what the result is if this is
+    /// happens or if the index is actually not in bounds.
+    pub fn in_bounds_index(&self, x: u32, y: u32, c: u8) -> usize {
+        self.format.in_bounds_index(x, y, c)
+    }
+
+    /// Shrink the image to the minimum of current and given extents.
+    ///
+    /// This does not modify the strides, so that the resulting sample buffer may have holes
+    /// created by the shrinking operation. Shrinking could also lead to an non-aliasing image when
+    /// samples had aliased each other before.
+    pub fn shrink_to(&mut self, width: u32, height: u32, channels: u8) {
+        self.format.shrink_to(width, height, channels)
+    }
+}
```

Comment:
Could FlatSamples also have a get() function implemented whenever Buffer: Index<usize>?

Reply:
I'm not sure how that would work exactly. For using the index kinds of the underlying buffer, I generally though `samples()[..]` to be good enough. If this is about `Index<(u32, u32)>`, sure.

Reply:
I was imagining something like:
```rust
fn get<T>(x: u32, y: u32, channel: u8) -> T
where
    Self::Buffer: Index<usize, T>,
    T: Copy
{
    self.buffer[self.format.index(x, y, channel).unwrap()]
}

Reply:
Ah, I see, I for some reason thought this was on `View`. `panic` makes sense due to behaviour of `ImageBuffer::get()` so yes, good idea.

Reply:
One odd thing is that the arguments of returning `T` instead of `&T` like `Vec::get` is not as strong as on `GenericImage`. That is, neither is the api old nor are we aggregating data.

Reply:
Added: 

```
impl<Buffer> Index<(u8, u32, u32)> for FlatSamples<Buffer>
    where Buffer: Index<usize>
{
    …
}
```

Reply:
And added mutable variants and more as well